### ISSUE TITLE
Type narrowing of string literal seems to get loosened in callbacks

### ISSUE THREAD

**Issue by Rycochet** [Microsoft/TypeScript#14748]
**TypeScript Version:**  2.2.1

**Code**

```ts
function myValueArg(arg: "value") { }

function myStringArg(arg: string) { }

namespace ns { // Not needed, but shows it's not scope related
    let str = "value";

    if (str === "value") {
        myValueArg(str);
        [].forEach(() => {
            myValueArg(str); // Argument of type 'string' is not assignable to parameter of type '"value"'.
        });
    } else {
        myStringArg(str);
        [].forEach(() => {
            myStringArg(str);
        });
    }
}
```

**Expected behavior:**
No warnings: the value (and type) of `str` has not changed.

**Actual behavior:**
Warning `Argument of type 'string' is not assignable to parameter of type '"value"'.`

**Comment by gcnew** [comment]
Use `const`. `let` bindings are reset because the function might be invoked asynchronously and a new value set to the `let` variable in the mean time.

**Comment by Rycochet** [comment]
That sort of makes sense (though this is just a sample, the real code I noticed it on is most definitely not a constant). Wouldn't it be better to annotate things that are definitely *not* asynchronous to prevent this spurious warning?

**Comment by andy-ms** [comment]
Duplicate of #9998.
